From 3e2c5d5adaa6f3e8dcb34e18f2799a937ed2d601 Mon Sep 17 00:00:00 2001
From: Alejandro Abdelnur <tucu@apache.org>
Date: Thu, 4 Oct 2012 23:55:26 +0000
Subject: [PATCH 1204/1357] MR1: MAPREDUCE-4556. FairScheduler: PoolSchedulable#updateDemand() has potential redundant computation (kkambatl via tucu)

git-svn-id: https://svn.apache.org/repos/asf/hadoop/common/branches/branch-1@1394331 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit ceb0b88f68b1da45a3ab854edecd0e040f5937bb)

Reason: FS Optimization
Ref: CDH-8454
Author: Karthik Kambatla
(cherry picked from commit f231e7eaa380a68c54f8ae287fbaa89767dd067b)
---
 .../org/apache/hadoop/mapred/PoolSchedulable.java  |   13 +++++++++----
 1 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/src/contrib/fairscheduler/src/java/org/apache/hadoop/mapred/PoolSchedulable.java b/src/contrib/fairscheduler/src/java/org/apache/hadoop/mapred/PoolSchedulable.java
index 251dfee..b782ecd 100644
--- a/src/contrib/fairscheduler/src/java/org/apache/hadoop/mapred/PoolSchedulable.java
+++ b/src/contrib/fairscheduler/src/java/org/apache/hadoop/mapred/PoolSchedulable.java
@@ -79,15 +79,20 @@ public class PoolSchedulable extends Schedulable {
    */
   @Override
   public void updateDemand() {
+    // limit the demand to maxTasks
+    int maxTasks = poolMgr.getMaxSlots(pool.getName(), taskType);
     demand = 0;
     for (JobSchedulable sched: jobScheds) {
       sched.updateDemand();
       demand += sched.getDemand();
+      if (demand >= maxTasks) {
+        demand = maxTasks;
+        break;
+      }
     }
-    // if demand exceeds the cap for this pool, limit to the max
-    int maxTasks = poolMgr.getMaxSlots(pool.getName(), taskType);
-    if(demand > maxTasks) {
-      demand = maxTasks;
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("The pool " + pool.getName() + " demand is " + demand
+          + "; maxTasks is " + maxTasks);
     }
   }
   
-- 
1.7.0.4

